Add support for signing during sync

jdieter opened the pull request:
This commit adds a new stage, RpmArtifactSigningStage, to synchronize.py that signs the rpms as they're being synchronized from an upstream repo. The new stage only does the signing if both a signing service and fingerprint have been set for the repo.

This is fully functional, but there are some questions I have on how best to move forward before moving this out of draft.

1. At the moment, signing will happen if the repo has a signing service and fingerprint, but this would be a behavior change with the potential of surprising users.  Should this instead be activated as a new sync policy or two ("additive_sign", "mirror_content_only_sign")?  Or should it be activated some other way?
2. How do we ensure that we're not resyncing packages each time a sync is run?  The checksums will obviously be different after redoing the signing, so should we add a field to the database that stores the original checksum?

I'm not sure if solving 2 is a hard requirement for this PR, but we really should decide on something for 1.

jdieter committed:
Initial support for signing during sync

This commit adds a new stage, RpmArtifactSigningStage, to synchronize.py
that signs the rpms as they're being synchronized from an upstream repo.
The new stage only does the signing if both a signing service and
fingerprint have been set for the repo.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Ensure signing only happens when using `*_sign` policies

This commit adds two new sync policies, `additive_sign` and
`mirror_content_only_sign`, which are identical to the non `_sign`
versions, but will sign packages during the sync process.

It also adds a test to ensure that signing *doesn't* happen when not using
a `*_sign` policy.

And, because sync is smart enough to skip packages that have identical
nevras, we need to ensure that all packages are garbage collected before
running the tests, so we're adding the delete_orphans_pre fixture to both
tests.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

pedro-psb reviewed (changes requested):
Sorry for not answering the question (1) and (2) before (you started working) and thanks for the contribution! We were in a rush for the last weeks.

I've commented on a few points already, feel free to ask for any clarification.
Also, I'd like to discuss your questions a bit more.

**About (1):**
I've heard (and agreed) a general preference of the team to define the "sign policy" in the repository level instead of a per-call basis. Would that work for you or is there a case where you would need per-call control over it?

If that works, there are some options. First, how do you think the "auto-sign" behavior (like we have in upload) would be a surprise? It's not something that can be configured by accident. Anyway, if we agree its a surprise, an alternative could be adding a `package_signing_enable` or similar to the repository, wdyt?

**About (2):**
I'm not convinced it won't download and sign it again... You could write a test for this, asserting that no new packages are added to the repository on a second sync.

It's worth mentioning that we have a recently opened issue for this (on the upload case) [here](https://github.com/pulp/pulp_rpm/issues/4045). Ideally we should share these kind of package signing handlers between upload and sync, but I don't really think we have a solution in place for this. I would be surprised if we do!

edit: IHO (2) is not a hard requirement for the PR.

jdieter committed:
Create a new artifact rather than updating old one

As mentioned in the PR review, it's far simpler to just create a new
artifact that handles all the hashing instead of manually creating the
hashes we think we should need.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Sign the content, not the artifact

We need to ensure that the size and pkgId are updated correctly, so pass
the content instead of the artifact to signing function and update the
content size and pkgId.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter commented on pulp_rpm/app/tasks/synchronizing.py:
```diff
@@ -1571,3 +1576,110 @@ def _handle_distribution_tree(declarative_content):
+    """
+    Stage for signing RPM artifacts during sync when repository has signing service configured.
+
+    This stage runs after ArtifactSaver, allowing us to sign the synchronized rpms
```
Yeah, dumb mistake (I was originally putting it after ArtifactDownloader, but discovered that that breaks things when the packages are stored anywhere except a filesystem).

jdieter replied:
Fixed

jdieter commented on pulp_rpm/app/tasks/synchronizing.py:
```diff
@@ -1571,3 +1576,110 @@ def _handle_distribution_tree(declarative_content):
+                if should_sign:
+                    for d_artifact in d_content.d_artifacts:
+                        await self._sign_rpm_artifact(d_artifact, signing_service, fingerprint)
+                await self.put(d_content)
```
I've implemented this now.

jdieter committed:
Move logic on whether or not to sign out of RpmArtifactSigningStage

As requested in the review, the logic on whether or not to do package
signing during sync has been moved out of RpmArtifactSigningStage, greatly
simplifying the stage.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Revert 'Ensure signing only happens when using `*_sign` policies'

As discussed in the review, adding new sync policies is the wrong place for
ensuring that signing happens only when necessary.  For the moment, we are
going to assume that if a user has set a signing service and fingerprint
for a repository, then they want it re-signed on sync.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

dralley commented on pulp_rpm/tests/functional/api/test_package_signing.py:
There are a few other cases where we explicitly fail if the user has chosen two incompatible behaviors - I think it's a bit of a judgement call but the consistent thing to do would be to refuse to allow the user to perform the sync.

dralley commented:
>     So, after using this a bit in our Pulp instance, I've figured out what was happening. It turns out that it won't download and sign again if and only if the repodata hasn't changed at all. If even one package has been updated in the repo, all the of the packages will be re-signed.

That makes sense, we do have an optimization to skip basically the whole sync process if the metadata checksum is unchanged from the previous sync.

> I think the correct solution is to store the fingerprint of any RPM signatures in the database, so we can then ensure that we don't re-sign if the content hash hasn't changed and the correct RPM signature is in the database. What do you think?

I assume what you mean is something along the lines of "keep a mapping of the pre-signed package checksums to the new, signed packages"?

jdieter committed:
During sync, skip package download for packages with unchanged NEVRAs

Currently, we start synchronizing packages even if the package is unchanged
from a previous sync.  This means that we end up re-signing packages each
time we sync when signing during sync.

This commit fixes that problem by re-using the already synced version of
the package for packages that already exist in the latest version of the
repository with the same NEVRA.  Note that this means that a package that
has been changed without changing the NEVRA will *not* be resynced anymore.
This may be seen as a security feature where packages can't be silently
changed.  To sync a package with the same NEVRA, the original package must
first be removed from the repository and then the repository must be synced
again.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
API will fail signed sync with incompatible policies

The "mirror_complete" sync policy or the "on_demand" remote policy are both
not supported when doing signing during syncing, so raise errors if either
policy is set during the sync process.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Improve test_optimize

Make it easier to detect when the repos aren't being populated as expected
by checking whether the repo_version has changed between each step of the
test.

This ensures that the test fails fast when things start going off the rails
rather than waiting until the end.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Properly handle DeclarativeContent when checking for duplicate files

We need more details in the DeclarativeContent, so let's create it properly
and only use it there's at least one artifact.

Also, add a "already_in_repo" flag into `extra_data` to ensure that we
don't try to re-sign packages that were already in the repository (and
should already be signed).

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter committed:
Add test to ensure that artifact downloading in _sign_rpm_content works

Create a test that syncs a repository, then syncs a completely different
repository, and then re-syncs the first.  Artifacts will be created during
the first sync that will be orphaned (since signing creates a new
artifact), and then, on the third sync, packages will be deduped to these
artifacts.  This test verifies that the artifact downloading in
_sign_rpm_content works with these deduped artifacts.

Signed-off-by: Jonathan Dieter <jdieter@ciq.com>

jdieter commented:
> The other option is to look only at the NEVRA, and if it is already present, don't sign, don't import. It's kind of a feature that old packages don't get sly-updated without extra work; broken packages would have to be deleted from the repository before a new sync to update them. This is a workaround I can get behind, because we wouldn't want things to update unexpectedly anyway.

Ok, this has been implemented (correctly, this time, I think).  Tests are all passing, and I've added a couple more, though I also had to disable two that were expecting the package to change when the NEVRA is the same.

I fine if we want to go down the route of adding a field for the original pkgId (and we'll probably also need to add a field for the signing key, so when looking through artifacts with the same originalPkgId, we can determine which signing key was used on each of them), but I wanted to make sure that the logic in here is sane and that what I've done isn't totally off-the-wall.

pedro-psb reviewed (commented):
Hello, sorry for taking so long.

I'm a bit reluctant about the `get_existing_package_nevras` caching part. There are edge cases where the NEVRA is the same but the content might have changed in the remote, so it doesnt feel safe to just use the local package. The `QueryExistingArtifacts`  stage should already prevent re-downloads by checksum. The main motivation was to avoid re-signing right? We now have a model to handle this (see [this PR](https://github.com/pulp/pulp_rpm/pull/4235/changes)), so I'll rather drop it for this PR. If there are other motivations, we can treat it a different issue/PR.

Having said that, the general approach for signing looks fine. You just need to update the signing stage to use the new mechanism for finding if a package is already signed with the given fingerprint (which the PR above provides). And it would be nice (we could do that later) to refactor some of `tasks/signing.py` utility functions to make them reusable here, which would help keeping the signing handling consistent.
